### PR TITLE
Add CounterUpDown implementation in the SDK

### PR DESCRIPTION
**Pull request by bogdandrutu**
This PR will be split into smaller PRs:
* #1202 
* #1200 
* #1208 
* #1210

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java [hunk 1]
```diff
@@ -185,6 +199,94 @@ public LongCounter build() {
     }
   }
 
+  /** No-op implementation of DoubleCounter interface. */
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java:202]:
```suggestion
  /** No-op implementation of DoubleUpDownCounter interface. */
```

File: api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java [hunk 2]
```diff
@@ -185,6 +199,94 @@ public LongCounter build() {
...
+      }
+
+      @Override
+      public DoubleUpDownCounter build() {
+        return new NoopDoubleUpDownCounter();
+      }
+    }
+  }
+
+  /** No-op implementation of CounterLong interface. */
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java:247]:
```suggestion
  /** No-op implementation of LongUpDownCounter interface. */
```

File: api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java [hunk 3]
```diff
@@ -185,6 +199,94 @@ public LongCounter build() {
...
+    @Override
+    public void add(long increment, String... labelKeyValuePairs) {}
+
+    @Override
+    public NoopBoundLongUpDownCounter bind(String... labelKeyValuePairs) {
+      Utils.validateLabelPairs(labelKeyValuePairs);
+      return NoopBoundLongUpDownCounter.INSTANCE;
+    }
+
+    /** No-op implementation of BoundLongCounter interface. */
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java:263]:
```suggestion
    /** No-op implementation of BoundLongUpDownCounter interface. */
```

File: api/src/main/java/io/opentelemetry/metrics/DoubleUpDownCounter.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+package io.opentelemetry.metrics;
+
+import io.opentelemetry.metrics.DoubleUpDownCounter.BoundDoubleUpDownCounter;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Counter is the most common synchronous instrument. This instrument supports an {@link
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/DoubleUpDownCounter.java:24]: All of this javadoc needs updating for the Up/Down.

File: api/src/main/java/io/opentelemetry/metrics/LongUpDownCounter.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+ *   private static final Meter meter = OpenTelemetry.getMeterRegistry().get("my_library_name");
+ *   private static final LongCounter counter =
+ *       meter.
+ *           .longCounterBuilder("processed_jobs")
+ *           .setDescription("Processed jobs")
+ *           .setUnit("1")
+ *           .build();
+ *
+ *   // It is recommended that the API user keep a reference to a Bound Counter.
+ *   private static final BoundLongCounter someWorkBound =
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/LongUpDownCounter.java:40]: The example in here needs to be updated.

File: api/src/test/java/io/opentelemetry/metrics/LongCounterTest.java [hunk 6]
```diff
@@ -86,6 +87,8 @@ public void noopBind_WithBadLabelSet() {
   public void doesNotThrow() {
```
Comment (jkwatson) [inline api/src/test/java/io/opentelemetry/metrics/LongCounterTest.java:87]: looks like we need a test that asserts that we can't increment with a negative value.

File: api/src/main/java/io/opentelemetry/metrics/SyncInstrument.java [hunk 7]
```diff
@@ -26,7 +26,7 @@
  * @since 0.3.0
  */
 @ThreadSafe
-public interface InstrumentWithBinding<B extends BoundInstrument> extends Instrument {
+public interface SyncInstrument<B extends BoundInstrument> extends Instrument {
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/metrics/SyncInstrument.java:29]: I think I'd prefer `SynchronousInstrument` here, even though it's a little longer. Also, I think some javadoc explaining what "synchronous" means in this context would be very useful.

File: api/src/test/java/io/opentelemetry/metrics/DoubleCounterTest.java [hunk 8]
```diff
@@ -87,6 +88,8 @@ public void noopBind_WithBadLabelSet() {
   public void doesNotThrow() {
```
Comment (jkwatson) [inline api/src/test/java/io/opentelemetry/metrics/DoubleCounterTest.java:88]: we should have a test to verify you can't pass in a negative increment

File: api/src/test/java/io/opentelemetry/metrics/DoubleUpDownCounterTest.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    thrown.expectMessage("key/value");
+    doubleUpDownCounter.bind("key");
+  }
+
+  @Test
+  public void doesNotThrow() {
+    DoubleUpDownCounter doubleUpDownCounter =
+        meter.doubleUpDownCounterBuilder(NAME).setDescription(DESCRIPTION).setUnit(UNIT).build();
+    BoundDoubleUpDownCounter bound = doubleUpDownCounter.bind();
+    bound.add(1.0);
```
Comment (jkwatson) [inline api/src/test/java/io/opentelemetry/metrics/DoubleUpDownCounterTest.java:92]: can we add a negative increment here, to validate it can actually go up and down?

**Comment by jkwatson** [comment]
This PR is huge. Can it be broken up a bit?

File: api/src/main/java/io/opentelemetry/metrics/SyncInstrument.java (hunk 7 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/metrics/SyncInstrument.java:29]: Done in #1202.

**Comment by bogdandrutu** [comment]
@jkwatson already started that, as you can see #1202 and #1200  are split from this. Will split more once these PRs are merged.

**Comment by jkwatson** [comment]
> @jkwatson already started that, as you can see #1202 and #1200 are split from this. Will split more once these PRs are merged.

I just noticed that.  :) I started reviewing in the wrong order.

**Comment by bogdandrutu** [comment]
@jkwatson: My bad that I did not mentioned it in the description, I updated description to make sure that is documented.

File: api/src/test/java/io/opentelemetry/metrics/DoubleCounterTest.java (hunk 8 above)
Comment (bogdandrutu) [reply api/src/test/java/io/opentelemetry/metrics/DoubleCounterTest.java]: Done.

File: api/src/test/java/io/opentelemetry/metrics/DoubleUpDownCounterTest.java (hunk 9 above)
Comment (bogdandrutu) [reply api/src/test/java/io/opentelemetry/metrics/DoubleUpDownCounterTest.java:92]: Done.

File: api/src/test/java/io/opentelemetry/metrics/LongCounterTest.java (hunk 6 above)
Comment (bogdandrutu) [reply api/src/test/java/io/opentelemetry/metrics/LongCounterTest.java]: Done.

File: api/src/main/java/io/opentelemetry/metrics/LongUpDownCounter.java (hunk 5 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/metrics/LongUpDownCounter.java]: Done.

File: api/src/main/java/io/opentelemetry/metrics/DoubleUpDownCounter.java (hunk 4 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/metrics/DoubleUpDownCounter.java]: Done.

**Comment by bogdandrutu** [comment]
@jkwatson this is now ready for review.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/LongCounterSdk.java [hunk 10]
```diff
@@ -69,7 +69,7 @@ BoundInstrument newBinding(Batcher batcher) {
     @Override
     public void add(long increment) {
       if (increment < 0) {
-        throw new IllegalArgumentException("monotonic counters can only increase");
+        throw new IllegalArgumentException("Counters can only increase");
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/LongCounterSdk.java:72]: with this throw, we should make sure to unbind above in a finally block.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java [hunk 11]
```diff
@@ -69,7 +69,7 @@ BoundInstrument newBinding(Batcher batcher) {
     @Override
     public void add(double increment) {
       if (increment < 0) {
-        throw new IllegalArgumentException("monotonic counters can only increase");
+        throw new IllegalArgumentException("Counters can only increase");
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java:72]: with this throw, we should make sure to unbind above in a finally block.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/LongCounterSdk.java (hunk 10 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/LongCounterSdk.java:75]: Done.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java (hunk 11 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java:75]: Done.
